[BUG] FString related compilation error in src/posix/i_steam.cpp

### GZDoom version

commit 6055ff029d5e87e84aaaf698bb430656e1e16588

### Which game are you running with GZDoom?

Other

### What Operating System are you using?

Linux x86_64

### Please describe your specific OS version

Debian GNU/Linux 12 (bookworm) x86_64

### Relevant hardware info

Intel i7-6700 / Intel HD Graphics 530

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

After latest `FString` related refactoring, compilation fails with:
```
/home/alex/prj/games/gzdoom/src/posix/i_steam.cpp: In function ‘TArray<FString> I_GetSteamPath()’:
/home/alex/prj/games/gzdoom/src/posix/i_steam.cpp:205:66: error: cannot convert ‘FString’ to ‘const char*’
  205 |                         SteamInstallFolders = ParseSteamRegistry(regPath);
      |                                                                  ^~~~~~~
      |                                                                  |
      |                                                                  FString
/home/alex/prj/games/gzdoom/src/posix/i_steam.cpp:119:55: note:   initializing argument 1 of ‘TArray<FString> ParseSteamRegistry(const char*)’
  119 | static TArray<FString> ParseSteamRegistry(const char* path)
      |                                           ~~~~~~~~~~~~^~~~
/home/alex/prj/games/gzdoom/src/posix/i_steam.cpp:224:38: error: cannot convert ‘FString’ to ‘const char*’
  224 |                         if(DirExists(candidate))
      |                                      ^~~~~~~~~
      |                                      |
      |                                      FString
In file included from /home/alex/prj/games/gzdoom/src/./doomtype.h:31,
                 from /home/alex/prj/games/gzdoom/src/./d_main.h:31,
                 from /home/alex/prj/games/gzdoom/src/posix/i_steam.cpp:42:
/home/alex/prj/games/gzdoom/src/common/utility/cmdlib.h:42:28: note:   initializing argument 1 of ‘bool DirExists(const char*)’
   42 | bool DirExists(const char *filename);
      |                ~~~~~~~~~~~~^~~~~~~~
make[2]: *** [src/CMakeFiles/zdoom.dir/build.make:112: src/CMakeFiles/zdoom.dir/posix/i_steam.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:739: src/CMakeFiles/zdoom.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

### Steps to reproduce the behaviour.

Explain how to reproduce
1. `cmake .. -DNO_FMOD=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=~/prj/games/ZMusic/build_install && make`
2. See above.

### Your configuration

_No response_

### Provide a Log

_No response_

--- a/src/common/platform/posix/cocoa/i_main.mm
+++ b/src/common/platform/posix/cocoa/i_main.mm
@@ -348,7 +348,7 @@ extern bool AppActive;
 
 	for (size_t i = 0, count = s_argv.Size(); i < count; ++i)
 	{
-		if (0 == strcmp(s_argv[i], charFileName))
+		if (0 == strcmp(s_argv[i].GetChars(), charFileName))
 		{
 			return FALSE;
 		}

--- a/src/common/platform/posix/osx/i_specialpaths.mm
+++ b/src/common/platform/posix/osx/i_specialpaths.mm
@@ -105,7 +105,7 @@ FString M_GetAppDataPath(bool create)
 	}
 
 	path += "/" GAMENAMELOWERCASE;
-	if (create) CreatePath(path);
+	if (create) CreatePath(path.GetChars());
 	return path;
 }
 
@@ -127,7 +127,7 @@ FString M_GetCachePath(bool create)
 	}
 
 	path += "/zdoom/cache";
-	if (create) CreatePath(path);
+	if (create) CreatePath(path.GetChars());
 	return path;
 }
 
@@ -169,9 +169,9 @@ FString M_GetConfigPath(bool for_reading)
 	{
 		// There seems to be no way to get Preferences path via NSFileManager
 		path += "/Preferences/";
-		CreatePath(path);
+		CreatePath(path.GetChars());
 
-		if (!DirExists(path))
+		if (!DirExists(path.GetChars()))
 		{
 			path = FString();
 		}
@@ -200,7 +200,7 @@ FString M_GetScreenshotsPath()
 	{
 		path += "/" GAME_DIR "/Screenshots/";
 	}
-	CreatePath(path);
+	CreatePath(path.GetChars());
 	return path;
 }
 
@@ -241,7 +241,7 @@ FString M_GetDocumentsPath()
 		path += "/" GAME_DIR "/";
 	}
 
-	CreatePath(path);
+	CreatePath(path.GetChars());
 	return path;
 }
 

--- a/src/common/platform/posix/osx/iwadpicker_cocoa.mm
+++ b/src/common/platform/posix/osx/iwadpicker_cocoa.mm
@@ -86,13 +86,13 @@ static const char* const tableHeaders[NUM_COLUMNS] = { "IWAD", "Game" };
 	for(int i = 0;i < numwads;i++)
 	{
 		NSMutableDictionary *record = [[NSMutableDictionary alloc] initWithCapacity:NUM_COLUMNS];
-		const char* filename = strrchr(wads[i].Path, '/');
+		const char* filename = strrchr(wads[i].Path.GetChars(), '/');
 		if(filename == NULL)
-			filename = wads[i].Path;
+			filename = wads[i].Path.GetChars();
 		else
 			filename++;
 		[record setObject:[NSString stringWithUTF8String:filename] forKey:[NSString stringWithUTF8String:tableHeaders[COLUMN_IWAD]]];
-		[record setObject:[NSString stringWithUTF8String:wads[i].Name] forKey:[NSString stringWithUTF8String:tableHeaders[COLUMN_GAME]]];
+		[record setObject:[NSString stringWithUTF8String:wads[i].Name.GetChars()] forKey:[NSString stringWithUTF8String:tableHeaders[COLUMN_GAME]]];
 		[data addObject:record];
 		[record release];
 	}
@@ -397,9 +397,9 @@ static void RestartWithParameters(const WadStuff& wad, NSString* parameters)
 		[arguments addObject:GetArchitectureString()];
 		[arguments addObject:executablePath];
 		[arguments addObject:@"-iwad"];
-		[arguments addObject:[NSString stringWithUTF8String:wad.Path]];
+		[arguments addObject:[NSString stringWithUTF8String:wad.Path.GetChars()]];
 		[arguments addObject:@"+defaultiwad"];
-		[arguments addObject:[NSString stringWithUTF8String:wad.Name]];
+		[arguments addObject:[NSString stringWithUTF8String:wad.Name.GetChars()]];
 		[arguments addObject:cvarArgument];
 
 		for (int i = 1, count = Args->NumArgs(); i < count; ++i)

--- a/src/common/platform/posix/sdl/i_system.cpp
+++ b/src/common/platform/posix/sdl/i_system.cpp
@@ -118,7 +118,7 @@ void Unix_I_FatalError(const char* errortext)
 		FString cmd;
 		cmd << "kdialog --title \"" GAMENAME " " << GetVersionString()
 			<< "\" --msgbox \"" << errortext << "\"";
-		popen(cmd, "r");
+		popen(cmd.GetChars(), "r");
 	}
 #ifndef NO_GTK
 	else if (I_GtkAvailable())
@@ -131,7 +131,7 @@ void Unix_I_FatalError(const char* errortext)
 		FString title;
 		title << GAMENAME " " << GetVersionString();
 
-		if (SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, title, errortext, NULL) < 0)
+		if (SDL_ShowSimpleMessageBox(SDL_MESSAGEBOX_ERROR, title.GetChars(), errortext, NULL) < 0)
 		{
 			printf("\n%s\n", errortext);
 		}
@@ -316,9 +316,9 @@ int I_PickIWad (WadStuff *wads, int numwads, bool showwin, int defaultiwad, int&
 
 		for(i = 0; i < numwads; ++i)
 		{
-			const char *filepart = strrchr(wads[i].Path, '/');
+			const char *filepart = strrchr(wads[i].Path.GetChars(), '/');
 			if(filepart == NULL)
-				filepart = wads[i].Path;
+				filepart = wads[i].Path.GetChars();
 			else
 				filepart++;
 			// Menu entries are specified in "tag" "item" pairs, where when a
@@ -329,9 +329,9 @@ int I_PickIWad (WadStuff *wads, int numwads, bool showwin, int defaultiwad, int&
 
 		if(defaultiwad >= 0 && defaultiwad < numwads)
 		{
-			const char *filepart = strrchr(wads[defaultiwad].Path, '/');
+			const char *filepart = strrchr(wads[defaultiwad].Path.GetChars(), '/');
 			if(filepart == NULL)
-				filepart = wads[defaultiwad].Path;
+				filepart = wads[defaultiwad].Path.GetChars();
 			else
 				filepart++;
 			cmd.AppendFormat(" --default \"%s (%s)\"", wads[defaultiwad].Name.GetChars(), filepart);
@@ -377,9 +377,9 @@ int I_PickIWad (WadStuff *wads, int numwads, bool showwin, int defaultiwad, int&
 	printf ("Please select a game wad (or 0 to exit):\n");
 	for (i = 0; i < numwads; ++i)
 	{
-		const char *filepart = strrchr (wads[i].Path, '/');
+		const char *filepart = strrchr (wads[i].Path.GetChars(), '/');
 		if (filepart == NULL)
-			filepart = wads[i].Path;
+			filepart = wads[i].Path.GetChars();
 		else
 			filepart++;
 		printf ("%d. %s (%s)\n", i+1, wads[i].Name.GetChars(), filepart);

--- a/src/common/platform/posix/sdl/i_system.cpp
+++ b/src/common/platform/posix/sdl/i_system.cpp
@@ -337,7 +337,7 @@ int I_PickIWad (WadStuff *wads, int numwads, bool showwin, int defaultiwad, int&
 			cmd.AppendFormat(" --default \"%s (%s)\"", wads[defaultiwad].Name.GetChars(), filepart);
 		}
 
-		FILE *f = popen(cmd, "r");
+		FILE *f = popen(cmd.GetChars(), "r");
 		if(f != NULL)
 		{
 			char gotstr[16];

--- a/src/common/platform/posix/sdl/sdlglvideo.cpp
+++ b/src/common/platform/posix/sdl/sdlglvideo.cpp
@@ -139,7 +139,7 @@ namespace Priv
 		caption.Format(GAMENAME " %s (%s)", GetVersionString(), GetGitTime());
 
 		const uint32_t windowFlags = (win_maximized ? SDL_WINDOW_MAXIMIZED : 0) | SDL_WINDOW_RESIZABLE | extraFlags;
-		Priv::window = SDL_CreateWindow(caption,
+		Priv::window = SDL_CreateWindow(caption.GetChars(),
 			(win_x <= 0) ? SDL_WINDOWPOS_CENTERED_DISPLAY(vid_adapter) : win_x,
 			(win_y <= 0) ? SDL_WINDOWPOS_CENTERED_DISPLAY(vid_adapter) : win_y,
 			win_w, win_h, windowFlags);
@@ -585,7 +585,7 @@ void I_SetWindowTitle(const char* caption)
 	{
 		FString default_caption;
 		default_caption.Format(GAMENAME " %s (%s)", GetVersionString(), GetGitTime());
-		SDL_SetWindowTitle(Priv::window, default_caption);
+		SDL_SetWindowTitle(Priv::window, default_caption.GetChars());
 	}
 }
 

--- a/src/common/platform/posix/unix/gtk_dialogs.cpp
+++ b/src/common/platform/posix/unix/gtk_dialogs.cpp
@@ -292,9 +292,9 @@ public:
 		store = gtk_list_store_new (3, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_INT);
 		for (int i = 0; i < numwads; ++i)
 		{
-			const char *filepart = strrchr (wads[i].Path, '/');
+			const char *filepart = strrchr (wads[i].Path.GetChars(), '/');
 			if (filepart == NULL)
-				filepart = wads[i].Path;
+				filepart = wads[i].Path.GetChars();
 			else
 				filepart++;
 			gtk_list_store_append (store, &iter);

--- a/src/common/platform/posix/unix/i_specialpaths.cpp
+++ b/src/common/platform/posix/unix/i_specialpaths.cpp
@@ -50,15 +50,15 @@ FString GetUserFile (const char *file)
 
 	path = NicePath("$HOME/" GAME_DIR "/");
 
-	if (stat (path, &info) == -1)
+	if (stat (path.GetChars(), &info) == -1)
 	{
 		struct stat extrainfo;
 
 		// Sanity check for $HOME/.config
 		FString configPath = NicePath("$HOME/.config/");
-		if (stat (configPath, &extrainfo) == -1)
+		if (stat (configPath.GetChars(), &extrainfo) == -1)
 		{
-			if (mkdir (configPath, S_IRUSR | S_IWUSR | S_IXUSR) == -1)
+			if (mkdir (configPath.GetChars(), S_IRUSR | S_IWUSR | S_IXUSR) == -1)
 			{
 				I_FatalError ("Failed to create $HOME/.config directory:\n%s", strerror(errno));
 			}
@@ -72,9 +72,9 @@ FString GetUserFile (const char *file)
 		// Transfer the old zdoom directory to the new location
 		bool moved = false;
 		FString oldpath = NicePath("$HOME/." GAMENAMELOWERCASE "/");
-		if (stat (oldpath, &extrainfo) != -1)
+		if (stat (oldpath.GetChars(), &extrainfo) != -1)
 		{
-			if (rename(oldpath, path) == -1)
+			if (rename(oldpath.GetChars(), path) == -1)
 			{
 				I_Error ("Failed to move old " GAMENAMELOWERCASE " directory (%s) to new location (%s).",
 					oldpath.GetChars(), path.GetChars());
@@ -115,7 +115,7 @@ FString M_GetAppDataPath(bool create)
 	FString path = NicePath("$HOME/.config/" GAMENAMELOWERCASE);
 	if (create)
 	{
-		CreatePath(path);
+		CreatePath(path.GetChars());
 	}
 	return path;
 }
@@ -135,7 +135,7 @@ FString M_GetCachePath(bool create)
 	FString path = NicePath("$HOME/.config/zdoom/cache");
 	if (create)
 	{
-		CreatePath(path);
+		CreatePath(path.GetChars());
 	}
 	return path;
 }
@@ -231,7 +231,7 @@ FString M_GetDemoPath()
 FString M_GetNormalizedPath(const char* path)
 {
 	char *actualpath;
-	actualpath = realpath(path, NULL);
+	actualpath = realpath(path.GetChars(), NULL);
 	if (!actualpath) // error ?
 		return nullptr;
 	FString fullpath = actualpath;

--- a/src/common/platform/posix/unix/i_specialpaths.cpp
+++ b/src/common/platform/posix/unix/i_specialpaths.cpp
@@ -74,7 +74,7 @@ FString GetUserFile (const char *file)
 		FString oldpath = NicePath("$HOME/." GAMENAMELOWERCASE "/");
 		if (stat (oldpath.GetChars(), &extrainfo) != -1)
 		{
-			if (rename(oldpath.GetChars(), path) == -1)
+			if (rename(oldpath.GetChars(), path.GetChars()) == -1)
 			{
 				I_Error ("Failed to move old " GAMENAMELOWERCASE " directory (%s) to new location (%s).",
 					oldpath.GetChars(), path.GetChars());
@@ -83,7 +83,7 @@ FString GetUserFile (const char *file)
 				moved = true;
 		}
 
-		if (!moved && mkdir (path, S_IRUSR | S_IWUSR | S_IXUSR) == -1)
+		if (!moved && mkdir (path.GetChars(), S_IRUSR | S_IWUSR | S_IXUSR) == -1)
 		{
 			I_FatalError ("Failed to create %s directory:\n%s",
 				path.GetChars(), strerror (errno));
@@ -231,7 +231,7 @@ FString M_GetDemoPath()
 FString M_GetNormalizedPath(const char* path)
 {
 	char *actualpath;
-	actualpath = realpath(path.GetChars(), NULL);
+	actualpath = realpath(path, NULL);
 	if (!actualpath) // error ?
 		return nullptr;
 	FString fullpath = actualpath;

--- a/src/common/utility/cmdlib.cpp
+++ b/src/common/utility/cmdlib.cpp
@@ -905,7 +905,7 @@ FString NicePath(const char *path)
 			slash = path + strlen(path);
 		}
 		FString who(path, slash - path);
-		pwstruct = getpwnam(who);
+		pwstruct = getpwnam(who.GetChars());
 	}
 	if (pwstruct == NULL)
 	{

--- a/src/ct_chat.cpp
+++ b/src/ct_chat.cpp
@@ -193,7 +193,7 @@ bool CT_Responder (event_t *ev)
 #ifdef __unix__
 		else if (ev->subtype == EV_GUI_MButtonDown)
 		{
-			CT_PasteChat(I_GetFromClipboard(true));
+			CT_PasteChat(I_GetFromClipboard(true).GetChars());
 		}
 #endif
 	}

--- a/src/d_main.cpp
+++ b/src/d_main.cpp
@@ -2041,7 +2041,7 @@ static void AddAutoloadFiles(const char *autoname, std::vector<std::string>& all
 
 #ifdef __unix__
 		file = NicePath("$HOME/" GAME_DIR "/skins");
-		D_AddDirectory (allwads, file, "*.wad", GameConfig);
+		D_AddDirectory (allwads, file.GetChars(), "*.wad", GameConfig);
 #endif	
 
 		// Add common (global) wads

--- a/src/gameconfigfile.cpp
+++ b/src/gameconfigfile.cpp
@@ -108,10 +108,10 @@ FGameConfigFile::FGameConfigFile ()
 		SetValueForKey ("Path", ".", true);
 		SetValueForKey ("Path", "$DOOMWADDIR", true);
 #ifdef __APPLE__
-		SetValueForKey ("Path", user_docs, true);
-		SetValueForKey ("Path", user_app_support, true);
+		SetValueForKey ("Path", user_docs.GetChars(), true);
+		SetValueForKey ("Path", user_app_support.GetChars(), true);
 		SetValueForKey ("Path", "$PROGDIR", true);
-		SetValueForKey ("Path", local_app_support, true);
+		SetValueForKey ("Path", local_app_support.GetChars(), true);
 #elif !defined(__unix__)
 		SetValueForKey ("Path", "$HOME", true);
 		SetValueForKey ("Path", "$PROGDIR", true);
@@ -133,10 +133,10 @@ FGameConfigFile::FGameConfigFile ()
 	{
 		SetSection ("FileSearch.Directories", true);
 #ifdef __APPLE__
-		SetValueForKey ("Path", user_docs, true);
-		SetValueForKey ("Path", user_app_support, true);
+		SetValueForKey ("Path", user_docs.GetChars(), true);
+		SetValueForKey ("Path", user_app_support.GetChars(), true);
 		SetValueForKey ("Path", "$PROGDIR", true);
-		SetValueForKey ("Path", local_app_support, true);
+		SetValueForKey ("Path", local_app_support.GetChars(), true);
 #elif !defined(__unix__)
 		SetValueForKey ("Path", "$PROGDIR", true);
 #else
@@ -156,14 +156,14 @@ FGameConfigFile::FGameConfigFile ()
 	{
 		SetSection("SoundfontSearch.Directories", true);
 #ifdef __APPLE__
-		SetValueForKey("Path", user_docs + "/soundfonts", true);
-		SetValueForKey("Path", user_docs + "/fm_banks", true);
-		SetValueForKey("Path", user_app_support + "/soundfonts", true);
-		SetValueForKey("Path", user_app_support + "/fm_banks", true);
+		SetValueForKey("Path", (user_docs + "/soundfonts").GetChars(), true);
+		SetValueForKey("Path", (user_docs + "/fm_banks").GetChars(), true);
+		SetValueForKey("Path", (user_app_support + "/soundfonts").GetChars(), true);
+		SetValueForKey("Path", (user_app_support + "/fm_banks").GetChars(), true);
 		SetValueForKey("Path", "$PROGDIR/soundfonts", true);
 		SetValueForKey("Path", "$PROGDIR/fm_banks", true);
-		SetValueForKey("Path", local_app_support + "/soundfonts", true);
-		SetValueForKey("Path", local_app_support + "/fm_banks", true);
+		SetValueForKey("Path", (local_app_support + "/soundfonts").GetChars(), true);
+		SetValueForKey("Path", (local_app_support + "/fm_banks").GetChars(), true);
 #elif !defined(__unix__)
 		SetValueForKey("Path", "$PROGDIR/soundfonts", true);
 		SetValueForKey("Path", "$PROGDIR/fm_banks", true);

--- a/src/playsim/actor.h
+++ b/src/playsim/actor.h
@@ -960,14 +960,14 @@ public:
 
 	double Distance2DSquared(AActor *other, bool absolute = false)
 	{
-		DVector2 otherpos = absolute ? other->Pos().XY() : other->PosRelative(this).XY();
-		return (Pos().XY() - otherpos).LengthSquared();
+		DVector3 otherpos = absolute ? other->Pos() : other->PosRelative(this);
+		return (Pos().XY() - otherpos.XY()).LengthSquared();
 	}
 
 	double Distance2D(AActor *other, bool absolute = false) const
 	{
-		DVector2 otherpos = absolute ? other->Pos().XY() : other->PosRelative(this).XY();
-		return (Pos().XY() - otherpos).Length();
+		DVector3 otherpos = absolute ? other->Pos() : other->PosRelative(this);
+		return (Pos().XY() - otherpos.XY()).Length();
 	}
 
 	double Distance2D(double x, double y) const
@@ -997,14 +997,14 @@ public:
 
 	DAngle AngleTo(AActor *other, bool absolute = false)
 	{
-		DVector2 otherpos = absolute ? other->Pos().XY() : other->PosRelative(this).XY();
-		return VecToAngle(otherpos - Pos().XY());
+		DVector3 otherpos = absolute ? other->Pos() : other->PosRelative(this);
+		return VecToAngle(otherpos.XY() - Pos().XY());
 	}
 
 	DAngle AngleTo(AActor *other, double oxofs, double oyofs, bool absolute = false) const
 	{
-		DVector2 otherpos = absolute ? other->Pos().XY() : other->PosRelative(this).XY();
-		return VecToAngle(otherpos - Pos() + DVector2(oxofs, oyofs));
+		DVector3 otherpos = absolute ? other->Pos() : other->PosRelative(this);
+		return VecToAngle(otherpos.XY() - Pos().XY() + DVector2(oxofs, oyofs));
 	}
 
 	DVector2 Vec2To(AActor *other) const

--- a/src/posix/i_steam.cpp
+++ b/src/posix/i_steam.cpp
@@ -202,7 +202,7 @@ TArray<FString> I_GetSteamPath()
 
 		try
 		{
-			SteamInstallFolders = ParseSteamRegistry(regPath);
+			SteamInstallFolders = ParseSteamRegistry(regPath.GetChars());
 		}
 		catch(class CRecoverableError &error)
 		{
@@ -221,7 +221,7 @@ TArray<FString> I_GetSteamPath()
 		{
 			struct stat st;
 			FString candidate(SteamInstallFolders[i] + "/" + AppInfo[app].BasePath);
-			if(DirExists(candidate))
+			if(DirExists(candidate.GetChars()))
 				result.Push(candidate);
 		}
 	}

--- a/src/posix/i_steam.cpp
+++ b/src/posix/i_steam.cpp
@@ -179,7 +179,7 @@ TArray<FString> I_GetSteamPath()
 	FString regPath = appSupportPath + "/Steam/config/config.vdf";
 	try
 	{
-		SteamInstallFolders = ParseSteamRegistry(regPath);
+		SteamInstallFolders = ParseSteamRegistry(regPath.GetChars());
 	}
 	catch(class CRecoverableError &error)
 	{

